File: PythonSpark/Preprocessing/PurchasesPreprocessingJob.py
from pyspark.sql import DataFrame

from Preprocessing.PreprocessingJob import PreprocessingJob
from Utils.Parsing import explode_data_frame_column, extract_array_element_to_new_column
from Utils.Repartitioning import repartition_data_frame_by_date
from Utils.TimeParsing import convert_unix_to_timestamp


class PurchasesPreprocessingJob(PreprocessingJob):

    def process(self, data: dict) -> DataFrame:
        time_fixed = convert_unix_to_timestamp(data["data"], "ts")
        purchases = self.__process_purchases__(time_fixed)
        return repartition_data_frame_by_date(purchases, "ts")

    @staticmethod
    def __process_purchases__(purchases: DataFrame) -> DataFrame:
        exploded_purchases = explode_data_frame_column(purchases, "items")
        purchases_with_item_extracted = extract_array_element_to_new_column(exploded_purchases, "items", 0, "item_id")
        purchases_with_amount_extracted = extract_array_element_to_new_column(purchases_with_item_extracted,
                                                                              "items", 1, "amount")
        return purchases_with_amount_extracted.drop("items")

File: ScalaSpark/src/scala/Preprocessing/PurchasesPreprocessingJob.scala
package Preprocessing

/**
 * This class prepares purchases for further processing (fixes time and explodes items)
 * Result is a dataframe with columns:
 *   user_id - id of a user, who made a purchase
 *   ip - ip address of a device where using which purchase was made
 *   ts - time when purchase was made
 *   item_id - id of an item purchased
 *   amount - amount of item with item_id purchased
 *   order_id - id of an order itself * @param arguments
 */
class PurchasesPreprocessingJob(arguments: Map[String, String], spark: SparkSession) extends PreprocessingJob(arguments, spark) {
  override def process(data: Map[String, DataFrame]): DataFrame = {
    val purchases = data("data")

    val timeFixed = convertUnixToTimeStamp("ts", purchases)

    val exploded = explodeDataFrameColumn("items", timeFixed)

    val itemsIdsExtracted = extractArrayElementToNewColumn("items", 0, "item_id", exploded)

    val amountsExtracted = extractArrayElementToNewColumn("items", 1, "amount", itemsIdsExtracted)

    val redundantColumnDropped = amountsExtracted.drop("items")

    repartitionByDate("ts", redundantColumnDropped)
  }
}

object PurchasesPreprocessingJob extends JobCompanion {
  def parser(): Parser = PreprocessingJob.parser()
}
